### PR TITLE
Add `.clang-format`, also add CLion build folder to the `.gitignore` file

### PR DESCRIPTION
**Pull request by kreuzerkrieg**
Initial attempt to add `.clang-format` which as close as possible to the current format in order to introduce as little noise as possible to the existing code. This format also follows [coding style guidelines](https://github.com/scylladb/seastar/blob/master/coding-style.md)

### CODE REVIEW

File: .clang-format
Comment (kreuzerkrieg) [inline .clang-format:88]: WDYT? is 140 a good value?

Comment (tchaikov) [reply .clang-format:88]: cc @lersek

File: .clang-format
Comment (tchaikov) [inline .clang-format:221]: we don't use Qt in scylla, so probably we don't need configurations like this? the same applies to Objective-C macros, and some other macros..

Comment (kreuzerkrieg) [reply .clang-format:221]: > we don't use Qt in scylla, so probably we don't need configurations like this? the same applies to Objective-C macros, and some other macros..

Do we have any macros that we have to treat as attribute/statement?

Comment (kreuzerkrieg) [reply .clang-format:221]: ok, then removing it

File: .clang-format
Comment (lersek) [reply .clang-format:88]: @tchaikov thanks for the CC

@kreuzerkrieg

> WDYT? is 140 a good value?

@avikivity recently suggested 160, so [that's what we have](https://github.com/scylladb/seastar/commit/bd27a45289ec13fddddcf383436cd6141a32c1b0) in the Seastar coding style now.

Comment (kreuzerkrieg) [reply .clang-format:88]: Your call, I can fit 250 characters on my display :). But it definitely can be a problem when working with laptop only.

**Comment by avikivity** [comment]
What's the impact of having .clang-format? What does it do?

**Comment by kreuzerkrieg** [comment]
> What's the impact of having .clang-format? What does it do?

Nothing as long as you dont use it. Otherwise it enforces that all team(s) use the same code format. In addition, it will reduce comments like [this one](https://github.com/scylladb/scylladb/pull/20120#discussion_r1714590126)
More on clang format
https://clang.llvm.org/docs/ClangFormat.html

**Comment by avikivity** [comment]
> > What's the impact of having .clang-format? What does it do?
> 
> Nothing as long as you dont use it. Otherwise it enforces that all team(s) use the same code format. In addition, it will reduce comments like [this one](https://github.com/scylladb/scylladb/pull/20120#discussion_r1714590126) More on clang format https://clang.llvm.org/docs/ClangFormat.html

I guess it makes sense to add clang-format-diff then.

**Comment by kreuzerkrieg** [comment]
> > > What's the impact of having .clang-format? What does it do?
> > 
> > 
> > Nothing as long as you dont use it. Otherwise it enforces that all team(s) use the same code format. In addition, it will reduce comments like [this one](https://github.com/scylladb/scylladb/pull/20120#discussion_r1714590126) More on clang format https://clang.llvm.org/docs/ClangFormat.html
> 
> I guess it makes sense to add clang-format-diff then.

As git hook? or as a part of CI validation?

In any case, baby steps

**Comment by tchaikov** [comment]
> > > > What's the impact of having .clang-format? What does it do?
> > > 
> > > 
> > > Nothing as long as you dont use it. Otherwise it enforces that all team(s) use the same code format. In addition, it will reduce comments like [this one](https://github.com/scylladb/scylladb/pull/20120#discussion_r1714590126) More on clang format https://clang.llvm.org/docs/ClangFormat.html
> > 
> > 
> > I guess it makes sense to add clang-format-diff then.
> 
> As git hook? or as a part of CI validation?
> 
> In any case, baby steps

i am also interested to see how well these rules fit our existing code style..  if they do not agree with existing code with a huge diff, we would need to adjust the rule to minimize the divergences. 

have you tried to run it against the tree and see how large the diff is?

**Comment by kreuzerkrieg** [comment]
> > > > > What's the impact of having .clang-format? What does it do?
> > > > 
> > > > 
> > > > Nothing as long as you dont use it. Otherwise it enforces that all team(s) use the same code format. In addition, it will reduce comments like [this one](https://github.com/scylladb/scylladb/pull/20120#discussion_r1714590126) More on clang format https://clang.llvm.org/docs/ClangFormat.html
> > > 
> > > 
> > > I guess it makes sense to add clang-format-diff then.
> > 
> > 
> > As git hook? or as a part of CI validation?
> > In any case, baby steps
> 
> i am also interested to see how well these rules fit our existing code style (if there is a fixed one =)).. if they do not agree with existing code with a huge diff, we would need to adjust the rule to minimize the divergences.
> 
> have you tried to run it against the tree and see how large the diff is?

not the whole tree, but I ran it on more than 15 files. and the result varies, some files left with only minor (like whitespaces) changes, some are heavily reformatted when the style used by the author strongly deviates from nowadays accepted style. But I dont think it matters, you dont have to reformat the whole file you are working on, reformatting only the code you added/touched is possible, well, depends on your IDE I guess

**Comment by regevran** [comment]
This is a great initiative by @kreuzerkrieg!
Personally I really want to see the code well formatted across the entire code-base. It has much more advantages than just "pretty code". 
However, getting there, especially in existing code bases, is a hard task.
I am not against baby-steps in getting into there - but I would like to see the plan that this baby step fits into.
I kindly ask @kreuzerkrieg  to come up with that plan before merging this PR.

**Review by regevran — COMMENTED** [review]
This is a great initiative by @kreuzerkrieg!
Personally I really want to see the code well formatted across the entire code-base. It has much more advantages than just "pretty code".
However, getting there, especially in existing code bases, is a hard task.
I am not against baby-steps in getting into there - but I would like to see the plan that this baby step fits into.
I kindly ask @kreuzerkrieg to come up with that plan before merging this PR.

**Comment by kreuzerkrieg** [comment]
> This is a great initiative by @kreuzerkrieg! Personally I really want to see the code well formatted across the entire code-base. It has much more advantages than just "pretty code". However, getting there, especially in existing code bases, is a hard task. I am not against baby-steps in getting into there - but I would like to see the plan that this baby step fits into. I kindly ask @kreuzerkrieg to come up with that plan before merging this PR.

I see it this way.
1) Add the `.clang-format`
2) Ask people to start using it, I guess objections regarding the style will arise :)
3) Get the format honed, so everyone (almost) is happy
4) (optional) drop in slack the git pre-commit hook that will format the committed code
5) Add format check to CI
Done

File: .clang-format
Comment (kreuzerkrieg) [reply .clang-format:88]: @avikivity so, any decision on it? 160?

**Review by lersek — CHANGES_REQUESTED** [review]
"Address PR comments" should not be a separate commit. The original series should be rebased, and each patch should be updated per prior review comments. In this case, that should be easy to do, as the series consists of a single patch, so any changes can be squashed (or the first commit can be update in-place with `git commit --amend`). Thanks!

**Comment by lersek** [comment]
@kreuzerkrieg

> 4\. drop in slack the git pre-commit hook that will format the committed code - it is up to the devs to start using it

I'd like to start using such a git pre-commit hook, but I don't think Slack is a good place to distribute the (recommended) hook. We could place the hook in `scripts/` instead; perhaps (and then it could be symlinked under `$GIT_DIR/hooks/` by whomever wanted to use it).

**Comment by kreuzerkrieg** [comment]
> @kreuzerkrieg
> 
> > 4. drop in slack the git pre-commit hook that will format the committed code - it is up to the devs to start using it
> 
> I'd like to start using such a git pre-commit hook, but I don't think Slack is a good place to distribute the (recommended) hook. We could place the hook in `scripts/` instead, perhaps (and then it could be symlinked under `$GIT_DIR/hooks/` by whomever wanted to use it).

Ok, will do the moment we decide we are pleased with existing format, already have it ready

**Comment by kreuzerkrieg** [comment]
> @kreuzerkrieg
> 
> > 4. drop in slack the git pre-commit hook that will format the committed code - it is up to the devs to start using it
> 
> I'd like to start using such a git pre-commit hook, but I don't think Slack is a good place to distribute the (recommended) hook. We could place the hook in `scripts/` instead, perhaps (and then it could be symlinked under `$GIT_DIR/hooks/` by whomever wanted to use it).

no problem, will do. BTW, we can configure the github to merge squashed PR to master so it will appear as one commit on master

**Comment by lersek** [comment]
@kreuzerkrieg

> we can configure the github to merge squashed PR to master so it will appear as one commit on master

In my (educated) opinion, squash-on-merge is a terrible workflow. It completely throws away an important aspect of contributions, namely the gradual and deliberate build-up of the final goal. The same cumulative diff can be achieved in many different ways, and precisely how a patch series is structured is relevant; it carries a lot of information. Code is primarily written for humans to read, thus, any patch set's main structuring principle is to guide the reader (assigned reviewer or later ad-hoc reader) through the construction of the feature or bugfix.

And then bisectability is another important reason for fine-grained (and well-isolated) patches.

**Comment by bitpathfinder** [comment]
@lersek 
> In my (educated) opinion, squash-on-merge is a terrible workflow. It completely throws away an important aspect of contributions, namely the gradual and deliberate build-up of the final goal. The same cumulative diff can be achieved in many different ways, and precisely how a patch series is structured is relevant; it carries a lot of information. Code is primarily written for humans to read, thus, any patch set's main structuring principle is to guide the reader (assigned reviewer or later ad-hoc reader) through the construction of the feature or bugfix.

There is an option to keep the original branches with all non-squashed commits, so in case a someone is interested in the history of developing a feature, they could refer to that branch.

**Comment by avikivity** [comment]
> @lersek
> 
> > In my (educated) opinion, squash-on-merge is a terrible workflow. It completely throws away an important aspect of contributions, namely the gradual and deliberate build-up of the final goal. The same cumulative diff can be achieved in many different ways, and precisely how a patch series is structured is relevant; it carries a lot of information. Code is primarily written for humans to read, thus, any patch set's main structuring principle is to guide the reader (assigned reviewer or later ad-hoc reader) through the construction of the feature or bugfix.
> 
> There is an option to keep the original branches with all non-squashed commits, so in case a someone is interested in the history of developing a feature, they could refer to that branch.

I'll keep my git log / blame / bisect thanks.

**Pull request closed without merging**
